src/posts/introducing-pint-xarray/index.md

```diff
@@ -0,0 +1,292 @@
+</p>
+
+We should take stories like this seriously: If we can automatically track units we can potentially eliminate a whole class of possible errors in our scientific work...
+
+## Pint tracks units
+
+There are a few packages for handling units in python (notably [unyt](https://github.com/yt-project/unyt) and [astropy.units](https://docs.astropy.org/en/stable/units/)), but for technical reasons we began units integration in Xarray with [pint](https://pint.readthedocs.io/en/stable/).
+These various packages work by providing a numerical array type that acts similarly to a NumPy array, and is intended to plug in and replace the raw NumPy array (a so-called "duck array type").
+
+Pint provides the `Quantity` object, which is a normal numpy array combined with a `pint.Unit`:
+
+```python
+q = np.array([6, 7]) * pint.Unit('metres')
+print(repr(q))
+```
+
+```
+Out: <Quantity([6 7], 'meter')>
+```
+
+Pint Quantities act like NumPy arrays, except that the units are carried around with the arrays, propagated through operations, and checked during operations involving multiple quantities.
+
+## Xarray now wraps Pint
+
+Thanks to the [tireless work](https://github.com/pydata/xarray/issues/3594) of xarray core developer Justus Magin, you can now enjoy this automatic unit-handling in xarray!
+
+Once you create a unit-aware Xarray object (see below for how) you can see the units of the data variables displayed as part of the printable representation.
+You also immediately get the key benefits of Pint:
+
+1. Units are propagated through arithmetic, and new quantities are built using the units of the inputs:
+
+   ```python
+   distance = xr.DataArray(10).pint.quantify("metres")
+   time = xr.DataArray(4).pint.quantify("seconds")
+
+   distance / time
+   ```
+
+   ```
+   Out:
+   <xarray.DataArray ()>
+   <Quantity(2.5, 'meter / second')>
+   ```
+
+2. Dimensionally inconsistent units are caught automatically:
+
+   ```python
+   apples = xr.DataArray(10).pint.quantify("kg")
+   oranges = xr.DataArray(200).pint.quantify("cm^3")
+
+   apples + oranges
+   ```
+
+   ```
+   Out:
+   DimensionalityError: Cannot convert from 'kilogram' ([mass]) to 'centimeter ** 3' ([length] ** 3)
+   ```
+
+3. Unit conversions become simple:
+
+   ```python
+   walk = xr.DataArray(500).pint.quantify('miles')
+
+   walk.pint.to('parsecs')
+   ```
+
+   ```
+   Out:
+   <xarray.DataArray ()>
+   <Quantity(2.6077643524162074e-11, 'parsec')>
+   ```
+
+4. You can specify that custom functions should expect certain units, and convert them if needed:
+
+   ```python
+   def jpl_trajectory_code(impulse):
+       """This function will only compute the correct result if supplied input in units of Newton-seconds."""
+
+       if impulse.pint.units != "newton * second":
+           impulse = impulse.pint.to("Newton * seconds")
+
+       # do some rocket science
+       ...
```

Comment:
This feels a bit strange to me: `pint` will happily do the computation, converting units along the way to make sure the results stay correct. At the end of the computation, all we have to do is ask to get the desired units. For this particular example:
```python
def jpl_trajectory_code(..., impulse):
    # do trajectory propagation
    ...

    return propagated_position.pint.to("km"), propagated_velocity.pint.to("km/s")
```

For functions that don't support units, we'd either write a wrapper that does the conversion (no need to check if the units already are what we need) and strips the units, then adds units to the result, or use `expects` once that's ready.

Reply:
Feel free to change the details of this example. The point just needs to be that unit-aware code would have prevented the error that befell the JPL mission. Exactly whether the code raises an error, converts, or whatever doesn't really matter for the pedagogical purposes of this blog post.

Reply:
I just noticed that with that it becomes essentially the same as 3 (unit conversion). Do you want to replace / extend that section? 4 should then be a section on wrapping library functions.

Reply:
Okay fair point that then they would be the same. Instead consider:

The point of section 3 is showing that you can do unit conversions, the point of section 4 is showing that you can require specific units for finicky library functions. The easiest way to demonstrate that you can require specific units on an input is to raise an error. We don't need to over-think it beyond that.

Therefore I now think that this would be fine for the purposes of this blog post

```python
   def jpl_trajectory_code(impulse):
       """A function which will only compute the correct result if supplied input in units of Newton-seconds."""

       if impulse.pint.units != "newton * second":
           raise pint.DimensionalityError()

       # do some rocket science
       ...
```
It gets the point across.

Yes in practice you would probably do this a bit differently, but that's what our pint-xarray docs should demonstrate.

Reply:
It feels a bit strange to have a blog post advertising `pint` and `pint-xarray` using unidiomatic code... I don't expect this blog post to become as often read as the one on [`kerchunk`](https://medium.com/pangeo/fake-it-until-you-make-it-reading-goes-netcdf4-data-on-aws-s3-as-zarr-for-rapid-data-access-61e33f8fe685), but that [had to be updated](https://github.com/fsspec/kerchunk/issues/170) because people kept using the code from the blog instead of from the documentation.

What if, instead of presenting it as a new feature, we'd show it as an example using 1 to 3? For unit-aware functions, that's what we'd actually end up doing.

Reply:
> What if, instead of presenting it as a new feature, we'd show it as an example using 1 to 3? 

Okay - what would that look like?

Reply:
the example would be the one from https://github.com/xarray-contrib/xarray.dev/pull/251#discussion_r940312605, but we'd describe it as something like:
```
With these, you can build code that automatically propagates units and converts them where necessary to stay consistent:

<code>

where the final explicit conversion is used to get the desired output units.
```
I'm sure we can improve the wording, though.

Reply:
@keewis I tried implementing your suggestion in https://github.com/xarray-contrib/xarray.dev/pull/251/commits/02838335e696a26edce21965e0fa82bdd70fb9e4, not quite sure if that's what you meant though.